pkg/kubectl/cmd/auth/cani.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	authorizationapi "k8s.io/kubernetes/pkg/apis/authorization"
+	internalauthorizationclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/authorization/internalversion"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
+// CanIOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type CanIOptions struct {
+	AllNamespaces bool
+	Quiet         bool
+	Namespace     string
+	SelfSARClient internalauthorizationclient.SelfSubjectAccessReviewsGetter
+
+	Verb         string
+	Resource     schema.GroupVersionResource
+	ResourceName string
+
+	Out io.Writer
+	Err io.Writer
+}
+
+func NewCmdCanI(f cmdutil.Factory, out, err io.Writer) *cobra.Command {
+	o := &CanIOptions{
+		Out: out,
+		Err: err,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "can-i VERB RESOURCE [NAME]",
+		Short: "check whether an action is allowed",
+		Long:  "check whether an action is allowed",
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(o.Complete(f, cmd, args))
+			cmdutil.CheckErr(o.Validate())
+
+			allowed, err := o.RunAccessCheck()
+			cmdutil.CheckErr(err)
+			if o.Quiet && !allowed {
+				os.Exit(2)
+			}
+		},
+	}
+
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().BoolVar(&o.AllNamespaces, "all-namespaces", o.AllNamespaces, "If true, check the specified action in all namespaces.")
+	cmd.Flags().BoolVarP(&o.Quiet, "quiet", "q", o.Quiet, "If true, suppress output and just return the exit code.")
+	return cmd
+}
+
+func (o *CanIOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []string) error {
+	switch len(args) {
+	case 3:
+		o.ResourceName = args[2]
+		fallthrough
+	case 2:
+		restMapper, _ := f.Object()
+		o.Verb = args[0]
+		o.Resource = resourceFor(restMapper, args[1])
+	default:
+		return errors.New("you must specify two or three arguments: verb, resource, and optional resourceName")
+	}
+
+	var err error
+	client, err := f.ClientSet()
```

Comment:
> IIUC, this will only work for core-apiservers and can't use user-apiservers directly. Is that fine?

Consider cases like a someone evaluating a new API server before turning it one.  He'll proxy to the service and then do things to check functionality and permissions.